File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/backup/BackupTask.java
package com.anod.car.home.prefs.backup;

import android.os.AsyncTask;

/**
 * @author alex
 * @date 12/30/13
 */
public class BackupTask extends AsyncTask<String, Void, Integer> {

    private int mType;

    private PreferencesBackupManager mBackupManager;

    private int mAppWidgetId;

    private BackupTaskListner mListener;

    public interface BackupTaskListner {

        void onBackupPreExecute(int type);

        void onBackupFinish(int type, int code);
    }

    public BackupTask(int type, PreferencesBackupManager backupManager, int appWidgetId,
            BackupTaskListner listener) {
        mType = type;
        mBackupManager = backupManager;
        mAppWidgetId = appWidgetId;
        mListener = listener;
    }

    @Override
    protected void onPreExecute() {
        mListener.onBackupPreExecute(mType);

    }

    protected Integer doInBackground(String... filenames) {
        String filename = filenames[0];
        if (filename == null) {
            return mBackupManager.doBackupInCar();
        }
        return mBackupManager.doBackupWidget(filename, mAppWidgetId);
    }

    protected void onPostExecute(Integer result) {
        mListener.onBackupFinish(mType, result);
    }
}

File: app/src/main/java/com/anod/car/home/backup/BackupTask.kt
package com.anod.car.home.backup

import android.content.ContentResolver
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * @author alex
 * @date 12/30/13
 */

class BackupTask(
        private val type: Int,
        private val backupManager: PreferencesBackupManager,
        private val appWidgetId: Int,
        private val uri: Uri) {

    constructor(backupManager: PreferencesBackupManager, appWidgetId: Int, uri: Uri)
            : this(PreferencesBackupManager.TYPE_MAIN, backupManager, appWidgetId, uri)

    constructor(backupManager: PreferencesBackupManager, uri: Uri)
            : this(PreferencesBackupManager.TYPE_INCAR, backupManager, 0, uri)


    interface BackupTaskListener {
        fun onBackupPreExecute(type: Int)
        fun onBackupFinish(type: Int, code: Int)
    }

    suspend fun execute(): Int = withContext(Dispatchers.IO) {

        if (type == PreferencesBackupManager.TYPE_INCAR) {
            if (ContentResolver.SCHEME_FILE == uri.scheme) {
                return@withContext backupManager.doBackupInCarLocal(uri)
            }
            return@withContext backupManager.doBackupInCarUri(uri)
        }

        return@withContext if (ContentResolver.SCHEME_FILE == uri.scheme) {
            backupManager.doBackupWidgetLocal(uri, appWidgetId)
        } else backupManager.doBackupWidgetUri(uri, appWidgetId)
    }
}
